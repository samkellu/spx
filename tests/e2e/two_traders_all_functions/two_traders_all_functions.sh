#! /bin/bash
RED='\033[0;31m'
GREEN='\033[0;32m'
YELLOW='\033[0;33m'
NC='\033[0m'

test_name="two_traders_all_functions"

gcc tests/e2e/$test_name/test_trader1.c -o tests/e2e/$test_name/test_trader1 -Wall -Werror -Wvla -O0 -std=c11 -g -fsanitize=address,leak -D_POSIX_C_SOURCE=199309L -lm
gcc tests/e2e/$test_name/test_trader2.c -o tests/e2e/$test_name/test_trader2 -Wall -Werror -Wvla -O0 -std=c11 -g -fsanitize=address,leak -D_POSIX_C_SOURCE=199309L -lm

./spx_exchange "products.txt" "tests/e2e/$test_name/test_trader1" "tests/e2e/$test_name/test_trader2" | tee tests/e2e/$test_name/$test_name.test

diff tests/e2e/$test_name/$test_name.out tests/e2e/$test_name/$test_name.test && echo -e "${GREEN} $test_name Exchange output correct.${NC}" || echo -e "${RED} $test_name Exchange output incorrect.${NC}"

diff tests/e2e/$test_name/trader1.out tests/e2e/$test_name/trader1.test && echo -e "${GREEN} $test_name Trader1 output correct.${NC}" || echo -e "${RED} $test_name Trader1 output incorrect.${NC}"
diff tests/e2e/$test_name/trader2.out tests/e2e/$test_name/trader2.test && echo -e "${GREEN} $test_name Trader2 output correct.${NC}" || echo -e "${RED} $test_name Trader2 output incorrect.${NC}"

rm tests/e2e/$test_name/$test_name.test
rm tests/e2e/$test_name/trader1.test
rm tests/e2e/$test_name/trader2.test
