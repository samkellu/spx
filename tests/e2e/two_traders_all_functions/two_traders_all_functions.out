[SPX] Starting
[SPX] Trading 2 products: GPU Router
[SPX] Created FIFO /tmp/spx_exchange_0
[SPX] Created FIFO /tmp/spx_trader_0
[SPX] Starting trader 0 (tests/e2e/two_traders_all_functions/test_trader1)
[SPX] Connected to /tmp/spx_exchange_0
[SPX] Connected to /tmp/spx_trader_0
[SPX] Created FIFO /tmp/spx_exchange_1
[SPX] Created FIFO /tmp/spx_trader_1
[SPX] Starting trader 1 (tests/e2e/two_traders_all_functions/test_trader2)
[SPX] Connected to /tmp/spx_exchange_1
[SPX] Connected to /tmp/spx_trader_1
[SPX] [T0] Parsing command: <SELL 0 GPU 30 500>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 0; Sell levels: 1
[SPX]		SELL 30 @ $500 (1 order)
[SPX]	Product: Router; Buy levels: 0; Sell levels: 0
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU 0 ($0), Router 0 ($0)
[SPX]	Trader 1: GPU 0 ($0), Router 0 ($0)
[SPX] [T1] Parsing command: <BUY 0 GPU 30 499>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 1; Sell levels: 1
[SPX]		SELL 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 0; Sell levels: 0
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU 0 ($0), Router 0 ($0)
[SPX]	Trader 1: GPU 0 ($0), Router 0 ($0)
[SPX] [T0] Parsing command: <SELL 1 GPU 30 501>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 1; Sell levels: 2
[SPX]		SELL 30 @ $501 (1 order)
[SPX]		SELL 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 0; Sell levels: 0
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU 0 ($0), Router 0 ($0)
[SPX]	Trader 1: GPU 0 ($0), Router 0 ($0)
[SPX] [T1] Parsing command: <BUY 1 GPU 30 500>
[SPX] Match: Order 0 [T0], New Order 1 [T1], value: $15000, fee: $150.
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 1; Sell levels: 1
[SPX]		SELL 30 @ $501 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 0; Sell levels: 0
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -30 ($15000), Router 0 ($0)
[SPX]	Trader 1: GPU 30 ($-15150), Router 0 ($0)
[SPX] [T0] Parsing command: <SELL 2 GPU 30 501>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 1; Sell levels: 1
[SPX]		SELL 60 @ $501 (2 orders)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 0; Sell levels: 0
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -30 ($15000), Router 0 ($0)
[SPX]	Trader 1: GPU 30 ($-15150), Router 0 ($0)
[SPX] [T1] Parsing command: <BUY 2 GPU 30 500>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 2; Sell levels: 1
[SPX]		SELL 60 @ $501 (2 orders)
[SPX]		BUY 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 0; Sell levels: 0
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -30 ($15000), Router 0 ($0)
[SPX]	Trader 1: GPU 30 ($-15150), Router 0 ($0)
[SPX] [T0] Parsing command: <SELL 3 GPU 30 502>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 2; Sell levels: 2
[SPX]		SELL 30 @ $502 (1 order)
[SPX]		SELL 60 @ $501 (2 orders)
[SPX]		BUY 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 0; Sell levels: 0
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -30 ($15000), Router 0 ($0)
[SPX]	Trader 1: GPU 30 ($-15150), Router 0 ($0)
[SPX] [T1] Parsing command: <BUY 3 GPU 30 501>
[SPX] Match: Order 1 [T0], New Order 3 [T1], value: $15030, fee: $150.
[SPX] Match: Order 2 [T0], New Order 3 [T1], value: $0, fee: $0.
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 2; Sell levels: 2
[SPX]		SELL 30 @ $502 (1 order)
[SPX]		SELL 30 @ $501 (1 order)
[SPX]		BUY 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 0; Sell levels: 0
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -60 ($30030), Router 0 ($0)
[SPX]	Trader 1: GPU 60 ($-30330), Router 0 ($0)
[SPX] [T0] Parsing command: <SELL 4 Router 22 51>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 2; Sell levels: 2
[SPX]		SELL 30 @ $502 (1 order)
[SPX]		SELL 30 @ $501 (1 order)
[SPX]		BUY 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 0; Sell levels: 1
[SPX]		SELL 22 @ $51 (1 order)
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -60 ($30030), Router 0 ($0)
[SPX]	Trader 1: GPU 60 ($-30330), Router 0 ($0)
[SPX] [T1] Parsing command: <BUY 4 Router 22 50>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 2; Sell levels: 2
[SPX]		SELL 30 @ $502 (1 order)
[SPX]		SELL 30 @ $501 (1 order)
[SPX]		BUY 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 1; Sell levels: 1
[SPX]		SELL 22 @ $51 (1 order)
[SPX]		BUY 22 @ $50 (1 order)
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -60 ($30030), Router 0 ($0)
[SPX]	Trader 1: GPU 60 ($-30330), Router 0 ($0)
[SPX] [T0] Parsing command: <CANCEL 3>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 2; Sell levels: 1
[SPX]		SELL 30 @ $501 (1 order)
[SPX]		BUY 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 1; Sell levels: 1
[SPX]		SELL 22 @ $51 (1 order)
[SPX]		BUY 22 @ $50 (1 order)
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -60 ($30030), Router 0 ($0)
[SPX]	Trader 1: GPU 60 ($-30330), Router 0 ($0)
[SPX] [T0] Parsing command: <AMEND 1 30 502>
[SPX] [T0] Parsing command: <AMEND 2 30 503>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 2; Sell levels: 1
[SPX]		SELL 30 @ $503 (1 order)
[SPX]		BUY 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 1; Sell levels: 1
[SPX]		SELL 22 @ $51 (1 order)
[SPX]		BUY 22 @ $50 (1 order)
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -60 ($30030), Router 0 ($0)
[SPX]	Trader 1: GPU 60 ($-30330), Router 0 ($0)
[SPX] [T1] Parsing command: <BUY 5 GPU 30 502>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 3; Sell levels: 1
[SPX]		SELL 30 @ $503 (1 order)
[SPX]		BUY 30 @ $502 (1 order)
[SPX]		BUY 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 1; Sell levels: 1
[SPX]		SELL 22 @ $51 (1 order)
[SPX]		BUY 22 @ $50 (1 order)
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -60 ($30030), Router 0 ($0)
[SPX]	Trader 1: GPU 60 ($-30330), Router 0 ($0)
[SPX] [T0] Parsing command: <AMEND 3 30 504>
[SPX] [T0] Parsing command: <AMEND 4 22 505>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 3; Sell levels: 1
[SPX]		SELL 30 @ $503 (1 order)
[SPX]		BUY 30 @ $502 (1 order)
[SPX]		BUY 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 1; Sell levels: 1
[SPX]		SELL 22 @ $505 (1 order)
[SPX]		BUY 22 @ $50 (1 order)
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -60 ($30030), Router 0 ($0)
[SPX]	Trader 1: GPU 60 ($-30330), Router 0 ($0)
[SPX] [T1] Parsing command: <BUY 6 Router 22 504>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 3; Sell levels: 1
[SPX]		SELL 30 @ $503 (1 order)
[SPX]		BUY 30 @ $502 (1 order)
[SPX]		BUY 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 2; Sell levels: 1
[SPX]		SELL 22 @ $505 (1 order)
[SPX]		BUY 22 @ $504 (1 order)
[SPX]		BUY 22 @ $50 (1 order)
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -60 ($30030), Router 0 ($0)
[SPX]	Trader 1: GPU 60 ($-30330), Router 0 ($0)
[SPX] [T0] Parsing command: <SELL 5 Router 100000 100000>
[SPX]	--ORDERBOOK--
[SPX]	Product: GPU; Buy levels: 3; Sell levels: 1
[SPX]		SELL 30 @ $503 (1 order)
[SPX]		BUY 30 @ $502 (1 order)
[SPX]		BUY 30 @ $500 (1 order)
[SPX]		BUY 30 @ $499 (1 order)
[SPX]	Product: Router; Buy levels: 2; Sell levels: 2
[SPX]		SELL 100000 @ $100000 (1 order)
[SPX]		SELL 22 @ $505 (1 order)
[SPX]		BUY 22 @ $504 (1 order)
[SPX]		BUY 22 @ $50 (1 order)
[SPX]	--POSITIONS--
[SPX]	Trader 0: GPU -60 ($30030), Router 0 ($0)
[SPX]	Trader 1: GPU 60 ($-30330), Router 0 ($0)
[SPX] Trader 1 disconnected
[SPX] Trader 0 disconnected
[SPX] Trading completed
[SPX] Exchange fees collected: $300
